test/mockupdb/test_handshake.py

```diff
@@ -11,14 +11,56 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-
-
-from mockupdb import MockupDB, OpReply, OpMsg, absent, Command, go
+import mockupdb
+from mockupdb import (MockupDB, OpReply, OpMsg, OpMsgReply, OpQuery, absent,
+                      Command, go, Request)
 from pymongo import MongoClient, version as pymongo_version
 from pymongo.errors import OperationFailure
+from pymongo.server_api import ServerApi
 
 import unittest
-
+from copy import deepcopy
+import time
+
+def test_handshake_with_option(self, protocol, num_servers, **kwargs):
+    primary = MockupDB(verbose=False)
+    secondaries = [MockupDB(verbose=False) for _ in range(num_servers-1)]
+    servers = [primary]+secondaries
+    hello = "ismaster" if isinstance(protocol(), OpQuery) else "hello"
+
+    # set up a custom handler to log and then immediately close all connections
+    # this makes sure that we only are checking the hello command, because
+    # after the first response it has been closed
+    reqs = []
+    def hangup(response):
+        reqs.append(deepcopy([response.doc, response.__class__]))
+        return response.hangup()
+    for server in servers:
+        server.autoresponds(hangup)
+        server.run()
+        self.addCleanup(server.stop)
+    hosts = [server.address_string for server in servers if not kwargs.get(
+        "loadBalanced") or server == primary]
+    address_str = "mongodb://"+','.join(hosts)
+    client = MongoClient(address_str,
+                         appname='my app',
+                         heartbeatFrequencyMS=500, # Speed up the test.
+                         **kwargs)
+    if isinstance(kwargs.get("server_api"), ServerApi):
+        kwargs.pop("server_api", None)
+        kwargs["apiVersion"] = "1"
+
+    self.addCleanup(client.close)
+    future = go(client.db.command, "whatever")
+    time.sleep(.1)
+    # check each server to see if it received messages, if it did then
+    # check them to make sure they match our protocol and kwargs
+    assert len(reqs) == len(hosts)
+    for i, t in reqs:
+        i = Request(i)
```

Comment:
Can we reuse the original request instead of creating a new one? Then we can reply on assert_matches to handle all the assertions, eg: `req.assert_matches(protocol(hello, **kwargs))